src/Vegetation/Canopy.jl

```diff
@@ -1,196 +1,643 @@
-
-The required parameters for the Medlyn stomatal conductance model.
-$(DocStringExtensions.FIELDS)
+struct CanopyModel{FT, RM, PM, SM, PHM, A, R, PS, D} <: AbstractModel{FT}
+    "Radiative transfer model, a canopy component model"
+    radiative_transfer::RM
+    "Photosynthesis model, a canopy component model"
+    photosynthesis::PM
+    "Stomatal conductance model, a canopy component model"
+    conductance::SM
+    "Plant hydraulics model, a canopy component model"
+    hydraulics::PHM
+    "Atmospheric forcing: prescribed or coupled"
+    atmos::A
+    "Radiative forcing: prescribed or coupled"
+    radiation::R
+    "Shared canopy parameters between component models"
+    parameters::PS
+    "Canopy model domain"
+    domain::D
+end
+
 """
-struct MedlynConductanceParameters{FT <: AbstractFloat}
-    "Relative diffusivity of water vapor (unitless)"
-    # TODO: move to CLIMAParameters"
-    Drel::FT
-    "Minimum stomatal conductance mol/m^2/s"
-    g0::FT
-    "Slope parameter, inversely proportional to the square root of marginal water use efficiency (Pa^{1/2})"
-    g1::FT
+    CanopyModel{FT}(;
+        radiative_transfer::AbstractRadiationModel{FT},
+        photosynthesis::AbstractPhotosynthesisModel{FT},
+        conductance::AbstractStomatalConductanceModel{FT},
+        hydraulics::AbstractPlantHydraulicsModel{FT},
+        atmos::AbstractAtmosphericDrivers{FT},
+        radiation::AbstractRadiativeDrivers{FT},
+        parameters::SharedCanopyParameters{FT, PSE},
+        domain::Union{
+            ClimaLSM.Domains.Point,
+            ClimaLSM.Domains.Plane,
+            ClimaLSM.Domains.SphericalSurface,
+        },
+    ) where {FT, PSE}
+
+An outer constructor for the `CanopyModel`. The primary
+constraints this applies are (1) ensuring that the domain is 1d or 2d
+(a ``surface" domain of a column, box, or sphere) and (2) ensuring
+consistency between the PlantHydraulics model and the general canopy model,
+since these are built separately.
+"""
+function CanopyModel{FT}(;
+    radiative_transfer::AbstractRadiationModel{FT},
+    photosynthesis::AbstractPhotosynthesisModel{FT},
+    conductance::AbstractStomatalConductanceModel{FT},
+    hydraulics::AbstractPlantHydraulicsModel{FT},
+    atmos::AbstractAtmosphericDrivers{FT},
+    radiation::AbstractRadiativeDrivers{FT},
+    parameters::SharedCanopyParameters{FT, PSE},
+    domain::Union{
+        ClimaLSM.Domains.Point,
+        ClimaLSM.Domains.Plane,
+        ClimaLSM.Domains.SphericalSurface,
+    },
+) where {FT, PSE}
+    if hydraulics.domain != domain
+        throw(
+            AssertionError(
+                "The provided canopy model domain must be the same as the hydraulics model domain.",
+            ),
+        )
+    end
+    if parameters.LAI != hydraulics.parameters.area_index[:leaf]
+        throw(
+            AssertionError(
+                "The leaf area index must be the same between the plant hydraulics and shared canopy parameters.",
```

Comment:
yes, we should. it ends up being a question of how much of a self contained model the plant hydraulics code is, vs something we only run with the canopy wrapper. since we are moving in that direction, probably the RHS of the plant hydraulics model should just grab the parameter from here